homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,406 @@
+
+from pyatv import exceptions, pair, scan
+from pyatv.const import Protocol
+from pyatv.convert import protocol_str
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_ADDRESS,
+    CONF_NAME,
+    CONF_PIN,
+    CONF_PROTOCOL,
+    CONF_TYPE,
+)
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import AbortFlow
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_CREDENTIALS, CONF_IDENTIFIER, CONF_START_OFF
+from .const import DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DEVICE_INPUT = "device_input"
+
+INPUT_PIN_SCHEMA = vol.Schema({vol.Required(CONF_PIN, default=None): int})
+
+DEFAULT_START_OFF = False
+PROTOCOL_PRIORITY = [Protocol.MRP, Protocol.DMAP, Protocol.AirPlay]
+
+
+async def device_scan(identifier, loop, cache=None):
+    """Scan for a specific device using identifier as filter."""
+
+    def _filter_device(dev):
+        if identifier is None:
+            return True
+        if identifier == str(dev.address):
+            return True
+        if identifier == dev.name:
+            return True
+        return any([service.identifier == identifier for service in dev.services])
+
+    def _host_filter():
+        try:
+            return [ip_address(identifier)]
+        except ValueError:
+            return None
+
+    if cache:
+        matches = [atv for atv in cache if _filter_device(atv)]
+        if matches:
+            return cache, matches[0]
+
+    for hosts in [_host_filter(), None]:
+        scan_result = await scan(loop, timeout=3, hosts=hosts)
+        matches = [atv for atv in scan_result if _filter_device(atv)]
+
+        if matches:
+            return scan_result, matches[0]
+
+    return scan_result, None
+
+
+def is_valid_credentials(credentials):
+    """Verify that credentials are valid for establishing a connection."""
+    return (
+        credentials.get(Protocol.MRP.value) is not None
+        or credentials.get(Protocol.DMAP.value) is not None
+    )
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_reauth(self, info):
+        """Handle initial step when updating invalid credentials."""
+        await self.async_set_unique_id(info[CONF_IDENTIFIER])
```

Comment:
We don't allow using ip addresses or other things that may change in the unique_id api. If there's no unique and stable id for the device we just have to check for existing identifiers before creating the entry, but can't use the unique_id api.

Reply:
Yes, I do know that and `CONF_IDENTIFIER` correspond to one of the unique identifiers of the device (as it can have several, see [docs](https://pyatv.dev/documentation/concepts/#identifiers)).

When calling from `__init__.py`:

```python
    def _auth_problem(self):
        ...
        identifier = self.config_entry.unique_id
        ...
        self.hass.async_create_task(
            self.hass.config_entries.flow.async_init(
                DOMAIN,
                context={"source": SOURCE_REAUTH},
                data={CONF_NAME: name, CONF_IDENTIFIER: identifier},
            )
        )
```

Reply:
If the identifier may not be an ip address, why do we ask the user for ip address?

> Start by entering the device name (e.g. Kitchen or Bedroom) or IP address of the Apple TV you want to add.

In the user step we're setting the device input as unique id.

Reply:
Right, after looking at this several times I now understand what you mean. The user is allowed to enter anything that can be used to find the expected device. But once scanning and filtering has been performed, the real unique is should be used. Would it perhaps make more sense if I set the unique id in the else branch (before moving on to the confirm step)? Here `self.atv.identifier` would be the true unique id.

Reply:
We want to set unique_id as early as possible in the flow, so we can abort as early as possible, but only using valid unique ids.

We should not use the unique_id attribute to store some arbitrary info that is later replaced with the correct unique id. If we need to store info in the flow for subsequent steps, save that info in a custom instance attribute initialized in the init method on the flow.

Reply:
The absolute earliest place where the unique id is known is in `async_find_device` and I guess it would be possible to set it there. But I also feel like... wouldn't I want to set it after the user has confirmed to add the device? Since the zeroconf discovery sets the unique id as well, that would prevent (I.e. abort due to trying to set the same unique id) manual adding a device that has been automatically discovered. Or is it not supposed to be possible to do that?

Reply:
We don't want to abort manual user flows in preference of discovery flows. Manual user flows should take precedence. The discovery flow should be aborted if a manual flow has been started with the same unique id.

In the user flow we should make sure to not raise on progress when setting the unique id:
https://github.com/home-assistant/core/blob/96035eb052fc6efa1ca5e5acc83d4a0d2fd6c168/homeassistant/config_entries.py#L903-L905

Reply:
Gotcha! I think I did the right thing, but I haven't tested it yet. Will try to do that soon. But I hope this is the last thing 🤞